python/lsst/meas/extensions/shapeHSM/_hsm_moments.py

```diff
@@ -0,0 +1,243 @@
+from lsst.geom import Point2I, Point2D
+
+
+class HsmMomentsConfig(SingleFramePluginConfig):
+    roundMoments = pexConfig.Field[bool](doc="Use round weight function?", default=False)
+    addFlux = pexConfig.Field[bool](doc="Store measured flux?", default=False)
+    subtractCenter = pexConfig.Field[bool](doc="Subtract starting center from x/y outputs?", default=False)
+
+
+class HsmMomentsPlugin(SingleFramePlugin):
+    ConfigClass = HsmMomentsConfig
+
+    def __init__(self, config, name, schema, logName=None):
+        super().__init__(config, name, schema, logName)
+
+        # Flag definitions.
+        flagDefs = FlagDefinitionList()
+        self.FAILURE = flagDefs.add("flag", "General failure flag, set if anything went wrong")
+        self.NO_PIXELS = flagDefs.add("flag_no_pixels", "No pixels to measure")
+        self.NOT_CONTAINED = flagDefs.add(
+            "flag_not_contained", "Center not contained in footprint bounding box"
+        )
+        self.PARENT_SOURCE = flagDefs.add("flag_parent_source", "Parent source, ignored")
+        self.GALSIM = flagDefs.add("flag_galsim", "GalSim failure")
+        self.EDGE = flagDefs.add("flag_edge", "Variance undefined outside image edge")
+
+        # Flag handler.
+        self.flagHandler = FlagHandler.addFields(schema, name, flagDefs)
+
+        # Safe centroid extractor.
+        self.centroidExtractor = measBase.SafeCentroidExtractor(schema, name)
+
+    @classmethod
+    def getExecutionOrder(cls):
+        return cls.SHAPE_ORDER
+
+    def fail(self, record, error=None):
+        # docstring inherited.
+        self.flagHandler.handleFailure(record)
+        if error:
+            center = record.getCentroid()
+            self.log.debug(
+                "Failed to measure shape for %d at (%f, %f): %s",
+                record.getId(),
+                center.getX(),
+                center.getY(),
+                error,
+            )
+
+
+class HsmSourceMomentsConfig(HsmMomentsConfig):
+    # later we can put some doscstring
+    badMaskPlanes = pexConfig.ListField[str](
+        doc="Mask planes used to reject bad pixels.", default=["BAD", "SAT"]
+    )
+
+
+@measBase.register("ext_shapeHSM_HsmSourceMoments")
+class HsmSourceMomentsPlugin(HsmMomentsPlugin):
+    ConfigClass = HsmSourceMomentsConfig
+
+    def __init__(self, config, name, schema, metadata, logName=None) -> None:
+        super().__init__(config, name, schema, logName)
+        # some flag thingy
+        # work in conjunction with fail()
+        self.centroidResultKey = afwTable.Point2DKey.addFields(
+            schema, name, "The centroid measured by HSM", "pixel"
+        )
+        self.shapeKey = afwTable.QuadrupoleKey.addFields(schema, name, doc="HSM source moments.")
+        if config.addFlux:  # search the correct attribute name in the codebase
+            self.fluxKey = measBase.FluxResultKey.addFields(schema, "flux", "The flux measured by HSM", "dn")
```

Comment:
```suggestion
            self.fluxKey = measBase.FluxResultKey.addFields(schema, name + "_Flux", "The flux measured by HSM", "dn")
```

Reply:
This entire `__init__` method should go to `HsmMomentsPlugin`.

Reply:
I realized the docs need to be class-specific. I kept them in individual classes for now, but we can return to this later.